src/gridfs/upload.rs

```diff
@@ -173,3 +179,287 @@ impl GridFsBucket {
+                    filename: stream.filename.take(),
+                    metadata: stream.metadata.take().unwrap(),
+                };
+
+                let new_future = close(stream.bucket.clone(), buffer, file).boxed();
+                stream.state.set_closing(new_future)
+            }
+            // This case is effectively unreachable, as the AsyncWriteExt methods take &mut self and
+            // poll the futures to completion. If a user were to call these polling methods directly
+            // and intersperse futures, we should just pend here until the writing future resolves.
+            State::Writing(_) => return Poll::Pending,
+            State::Closing(future) => future,
+            State::Closed => return Poll::Ready(Err(get_closed_error())),
+        };
+
+        let result = match future.poll_unpin(cx) {
+            Poll::Ready(result) => result,
+            Poll::Pending => return Poll::Pending,
+        };
+
+        stream.state = State::Closed;
+        match result {
+            Ok(()) => Poll::Ready(Ok(())),
+            Err(error) => Poll::Ready(Err(error.into_futures_io_error())),
+        }
+    }
+}
+
+async fn write_bytes(
+    bucket: GridFsBucket,
+    mut buffer: Vec<u8>,
+    starting_n: u32,
+    chunk_size: u32,
+    files_id: Bson,
+) -> Result<(u32, Vec<u8>)> {
+    bucket.create_indexes().await?;
+
+    let mut n = 0;
+    let mut chunks = vec![];
+
+    while buffer.len() as u32 - (n * chunk_size) >= chunk_size {
+        let start = n * chunk_size;
+        let end = (n + 1) * chunk_size;
+        let chunk = Chunk {
+            id: ObjectId::new(),
+            files_id: files_id.clone(),
+            n: starting_n + n,
+            data: RawBinaryRef {
+                subtype: BinarySubtype::Generic,
+                bytes: &buffer[(start as usize)..(end as usize)],
+            },
+        };
+        n += 1;
+        chunks.push(chunk);
+    }
+
+    bucket.chunks().insert_many(chunks, None).await?;
+    buffer.drain(..(n * chunk_size) as usize);
+
+    Ok((n, buffer))
+}
+
+async fn close(bucket: GridFsBucket, buffer: Vec<u8>, file: FilesCollectionDocument) -> Result<()> {
+    let insert_result: Result<()> = async {
+        if !buffer.is_empty() {
+            let final_chunk = Chunk {
+                id: ObjectId::new(),
+                n: file.n() - 1,
+                files_id: file.id.clone(),
+                data: RawBinaryRef {
+                    subtype: BinarySubtype::Generic,
+                    bytes: &buffer[..],
+                },
+            };
+            bucket.chunks().insert_one(final_chunk, None).await?;
+        }
+        bucket.files().insert_one(&file, None).await?;
+        Ok(())
+    }
+    .await;
+
+    match insert_result {
+        Ok(()) => Ok(()),
+        Err(error) => {
+            bucket
+                .chunks()
+                .delete_many(doc! { "files_id": file.id }, None)
+                .await?;
```

Comment:
Hm this is kind of a weird case here if we run into an error executing the `delete_many`. I wonder if it would be better to issue a `GridFs` error that was something like `"failed inserting files document: {}, then failed removing orphaned chunks: {}"`. WDYT?